src/VisualStudio/Core/Def/Implementation/Library/FindResults/LibraryManager_FindReferences.cs

```diff
@@ -1,187 +1,100 @@
-        /// definitions are displayed in Find Symbol Results, it only changes which definition
-        /// a given reference should appear under when its location is a reference to multiple
-        /// definitions.
-        /// </summary>
-        private int GetDefinitionPrecedence(ReferencedSymbol referencedSymbol)
+        private ImmutableArray<DefinitionTreeItem> CreateDefinitionItems(
+            DefinitionItem definitionItem,
+            DefinitionsAndReferences definitionsAndReferences,
+            int commonPathElements)
         {
-            switch (referencedSymbol.Definition.Kind)
-            {
-                case SymbolKind.Event:
-                case SymbolKind.Field:
-                case SymbolKind.Label:
-                case SymbolKind.Local:
-                case SymbolKind.Method:
-                case SymbolKind.Parameter:
-                case SymbolKind.Property:
-                case SymbolKind.RangeVariable:
-                    return 0;
-
-                case SymbolKind.ArrayType:
-                case SymbolKind.DynamicType:
-                case SymbolKind.ErrorType:
-                case SymbolKind.NamedType:
-                case SymbolKind.PointerType:
-                    return 1;
-
-                default:
-                    return 2;
-            }
+            var referenceItems = CreateReferenceItems(
+                definitionItem, definitionsAndReferences, commonPathElements);
+
+            return ConvertToDefinitionTreeItems(definitionItem, referenceItems);
         }
 
-        private AbstractTreeItem ConvertToDefinitionItem(
-            Solution solution,
-            ReferencedSymbol referencedSymbol,
-            Location location,
-            Glyph glyph)
+        private ImmutableArray<DefinitionTreeItem> ConvertToDefinitionTreeItems(
+            DefinitionItem definitionItem,
+            ImmutableArray<SourceReferenceTreeItem> referenceItems)
         {
-            if (!location.IsInSource)
-            {
-                return referencedSymbol.Locations.Any()
-                    ? new MetadataDefinitionTreeItem(
-                        solution.Workspace,
-                        referencedSymbol.Definition,
-                        referencedSymbol.Locations.First().Document.Project.Id,
-                        glyph.GetGlyphIndex())
-                    : null;
-            }
+            var result = ImmutableArray.CreateBuilder<DefinitionTreeItem>();
 
-            var document = solution.GetDocument(location.SourceTree);
-            var sourceSpan = location.SourceSpan;
-            if (!IsValidSourceLocation(document, sourceSpan))
+            for (int i = 0, n = definitionItem.Locations.Length; i < n; i++)
             {
-                return null;
+                var location = definitionItem.Locations[i];
+
+                // Each definition item may end up as several top nodes (because of partials).
```

Comment:
I'd much rather focus on just keeping the existing synchronous FAR working the exact same way as today, and instead making hte new FAR have the better experience.  Every bit of effort i spend improving the current far is less time to make the new FAR as good as possible :-/